hub/integrations/pytorch.py

```diff
@@ -53,19 +71,34 @@ def dataset_to_pytorch(dataset, transform: Callable = None, workers: int = 1):
 
 class TorchDataset:
     def __init__(self, dataset, transform: Callable = None, workers: int = 1):
-        self.dataset = dataset
         self._set_globals()
         self.transform: Optional[Callable] = transform
         self.workers: int = workers
         self.map = ProcessPool(nodes=workers).map
-        self.len = len(dataset)
-        self.keys = list(self.dataset.tensors)
+        self.length = len(dataset)
+        self.keys = list(dataset.tensors)
+        self.storage = remove_memory_cache(dataset.storage)
+
+        if isinstance(self.storage, S3Provider):
+            self.storage_state_tuple = self.storage.__getstate__()
 
         # contains meta for each Tensor
         self.all_tensor_metas: Dict[str, TensorMeta] = self._load_all_meta()
+        index_value = dataset.index.values[0].value
+        if not isinstance(index_value, slice):
+            raise DatasetUnsupportedPytorch(
+                "Only full dataset or dataset indexed using slices can be converted to pytorch."
+            )
+
+        if index_value.step not in [None, 1]:
+            raise DatasetUnsupportedPytorch(
+                "The step of the Dataset object is not None or 1"
+            )
+
+        self.index_offset = index_value.start or 0
 
         # contains index_meta for each Tensor
-        self.all_index_metas: Dict[str, IndexMeta] = self._load_all_index_metas()
+        self.all_index_metas: Dict[str, IndexMeta] = self._load_all_index_meta()
```

Comment:
to be honest, slightly unfortunate we are using index meta without dataset class here and directly parsing chunks. I know it adds speed, but would have better to make sure core works faster and everyone use core.

Reply:
We aren't directly parsing chunks, we're using the Index meta class. The current way to do this through the api would be to slice the dataset to get every tensor and get index metas from that, which isn't really the best way I think